backend.go

```diff
@@ -800,91 +808,172 @@ type DecodeData struct {
 // 	}
 // }
 
-// DecodeMatrix decode all the subchunk of frags, and linearize data
-func (backend *Backend) DecodeMatrix(frags [][]byte, piecesize int) (*DecodeData, error) {
+type RawFragment = []byte
+type ValidatedFragment = []byte
+
+func (backend *Backend) ValidateFragmentMatrix(frag RawFragment, pieceSize int) bool {
+	result := C.check_matrix_fragment((*C.char)(unsafe.Pointer(&frag[0])), C.int(len(frag)), C.int(pieceSize))
+	if result != C.bool(true) {
+		return false
+	}
+
+	return true
+}
+
+type ChunkInfo struct {
+	ChunkSize int
+	NrChunk   int
+}
+
+func (backend *Backend) ChunkInfo(fragRangeLen int, pieceSize int) ChunkInfo {
+	chunkSize := pieceSize + backend.headerSize
+	nrChunks := fragRangeLen / chunkSize
+	if nrChunks*chunkSize != fragRangeLen {
+		nrChunks++
+	}
+
+	return ChunkInfo{
+		ChunkSize: chunkSize,
+		NrChunk:   nrChunks,
+	}
+}
+
+func (backend *Backend) LinearizeMatrix(frags []ValidatedFragment, pieceSize int) (*DecodeData, error) {
 	var wg sync.WaitGroup
 
 	if len(frags) == 0 {
-		return nil, errors.New("decoding requires at least one fragment")
+		return nil, errors.New("linearizing requires at least one fragment")
 	}
 
-	fragLen := len(frags[0])
-	lenBlock := piecesize + backend.headerSize
-	numBlock := fragLen / lenBlock
-	if numBlock*lenBlock != fragLen {
-		numBlock++
+	fragRangeLen := len(frags[0])
+	chunkInfo := backend.ChunkInfo(fragRangeLen, pieceSize)
+
+	/* Fragments are sorted beforehand with the index of the first chunk.
+	   All chunks of a fragments share the same index. */
+	fragsIndex := make([]int, len(frags))
+	for i := 0; i < len(frags); i += 1 {
+		fragsIndex[i] = i
 	}
 
-	// allocate output buffer
-	dataB, data := backend.pool.New(numBlock * piecesize * backend.K)
+	sort.Slice(fragsIndex, func(i, j int) bool {
+		lhs := frags[fragsIndex[i]]
+		rhs := frags[fragsIndex[j]]
 
+		var lhsIdx, rhsIdx C.int
+		lhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&lhs[0])))
+		rhsIdx = C.get_fragment_idx((*C.char)(unsafe.Pointer(&rhs[0])))
+
+		return lhsIdx < rhsIdx
+	})
+
+	/* There is no reconstruction that can happen.
+	   All coding fragments must be ignored */
+	lastDataFragIdxExcl := 0
+	previousFragIdx := -1
+	for lastDataFragIdxExcl < len(fragsIndex) {
+		frag := frags[fragsIndex[lastDataFragIdxExcl]]
+		idx := int(C.get_fragment_idx((*C.char)(unsafe.Pointer(&frag[0]))))
+
+		if idx < 0 {
+			return nil, errors.New("invalid fragment index")
+		}
+
+		if idx >= backend.K {
```

Comment:
This check might lead to an unexpected case as the gap check is done after this check.
e.g. for k=2, n=1, the fragments index are: 0, 2 => `lastDataFragIdxExcl=1`